In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [2]:
brk = pd.read_csv('BRK13F.csv', header = 0, index_col = 0)/100
brk.index = pd.to_datetime(brk.index).to_period('M')

In [3]:
brk['BRK-RF'] = brk['BRK.A']-brk['RF']
brk['MKT-RF'] = brk['MKT']-brk['RF']

In [4]:
brk

BRK.A     MKT     SMB     HML       RF     QMJ     BAB     UMD  \
datadate                                                                    
1999-01   0.0201  0.0363 -0.0124 -0.0361  0.00364 -0.0354 -0.0387  0.0234   
1999-02  -0.0153 -0.0413 -0.0385  0.0215  0.00379 -0.0041  0.0094 -0.0040   
1999-03   0.0150  0.0352 -0.0385 -0.0298  0.00364 -0.0250 -0.0773 -0.0251   
1999-04   0.0533  0.0456  0.0288  0.0347  0.00369 -0.0454 -0.0020 -0.0913   
1999-05  -0.0207 -0.0255  0.0564  0.0115  0.00378  0.0182  0.0362 -0.0298   
...          ...     ...     ...     ...      ...     ...     ...     ...   
2013-08  -0.0406 -0.0260 -0.0003 -0.0177  0.00003  0.0068 -0.0031 -0.0031   
2013-09   0.0224  0.0376  0.0242 -0.0114  0.00002 -0.0078  0.0089  0.0289   
2013-10   0.0355  0.0407 -0.0151  0.0053  0.00003  0.0184  0.0287  0.0068   
2013-11   0.0337  0.0265  0.0118 -0.0062  0.00005  0.0143  0.0189  0.0188   
2013-12   0.0335  0.0272 -0.0015  0.0009  0.00006 -0.0159  0.0043  0.0110   

           BRK-RF   MKT-RF  
datadate                    
1999-01   0.01646  0.03266  
1999-02  -0.01909 -0.04509  
1999-03   0.01136  0.03156  
1999-04   0.04961  0.04191  
1999-05  -0.02448 -0.02928  
...           ...      ...  
2013-08  -0.04063 -0.02603  
2013-09   0.02238  0.03758  
2013-10   0.03547  0.04067  
2013-11   0.03365  0.02645  
2013-12   0.03344  0.02714  

[180 rows x 10 columns]

In [5]:
brk.describe()

BRK.A         MKT         SMB         HML          RF         QMJ  \
count  180.000000  180.000000  180.000000  180.000000  180.000000  180.000000   
mean     0.006692    0.003991    0.004575    0.002722    0.001770    0.003384   
std      0.044561    0.047667    0.029171    0.030461    0.001679    0.034059   
min     -0.143700   -0.185500   -0.075700   -0.088100    0.000010   -0.104500   
25%     -0.011500   -0.023425   -0.013175   -0.013450    0.000120   -0.014200   
50%      0.010450    0.010300    0.002800    0.001650    0.001345   -0.000450   
75%      0.031825    0.036350    0.020850    0.019650    0.003535    0.021575   
max      0.152600    0.116300    0.121900    0.133900    0.005160    0.128900   

              BAB         UMD      BRK-RF      MKT-RF  
count  180.000000  180.000000  180.000000  180.000000  
mean     0.005086    0.003679    0.004922    0.002220  
std      0.044355    0.057949    0.044693    0.047843  
min     -0.152200   -0.347600   -0.147650   -0.185870  
25%     -0.010750   -0.012825   -0.014800   -0.025310  
50%      0.008400    0.008450    0.008775    0.009550  
75%      0.025225    0.029100    0.029483    0.032935  
max      0.129000    0.170900    0.148460    0.116290

## Problem 1-1

Case.1
$$BRK-RF = \alpha + \beta_{MKT}(RM-RF) + \beta_{SMB}(SMB) + \beta_{HML}(HML) +\epsilon$$

In [17]:
Y11 = brk['BRK-RF']
X11 = brk[['MKT-RF', 'SMB', 'HML']]
X11 = sm.add_constant(X11, has_constant = 'add')
model = sm.OLS(Y11,X11).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 BRK-RF   R-squared:                       0.620
Model:                            OLS   Adj. R-squared:                  0.614
Method:                 Least Squares   F-statistic:                     95.92
Date:                Thu, 09 Mar 2023   Prob (F-statistic):           7.88e-37
Time:                        12:32:11   Log-Likelihood:                 391.72
No. Observations:                 180   AIC:                            -775.4
Df Residuals:                     176   BIC:                            -762.7
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0033      0.002      1.544      0.1

1. No. Obsevations : 'Number of obsevations"으로 총 표본 수  
>case) 1999-01 ~ 2013-12 기간(180일) 동안 각 컬럼의 데이터 수

2. DF Residuals : 잔차의 자유도  
전체 표본 수에서 측정되는 변수들(종속변수 및 독립변수)의 개수를 뺸것.    
cf. 자유도 : 조정된 표본, 표분분산의 평균이 모분산이 되도록 하기 위한 개념
>case) 180 - (3 + 1) = 176, 180개의 표본 중 측정되는 변수(제약조건 4개)를 제외 한 176개로 잔차의 분산을 구한다.  

3. DF Models : 독립변수의 개수

4. R-squared : 결정계수  
최소제곱법(MLS, Method of Least Squares)을 활용한 선형 회귀분석에서 종속변수의 분산(SST) 중에서 추정값(회귀값)의 분산(SSR)의 비율을 의미한다. -> 회귀 변동이 실제 변동(분산)을 얼마만큼 포착하는가를 의미한다. $$R^2 = \frac{SSR}{SST} = \frac{\textstyle\sum(\hat{y_i} - \overline{y})^2}{\textstyle\sum(y_i - \overline{y})^2}$$
>case) $R^2$ = 0.62, 실제 분산이 1이라고 했을 때 추정값(회귀값)의 분산은 0.62, 실제 변동의 62%를 포착했다(설명할 수 있다). 


5. Adj. R-squared : 수정 결정계수  
* 결정계수의 한계  
결정계수는 독립변수 개수가 많아질 수록 그 값이 커지게 된다. 종속변수의 변동을 설명해 주지 못하는 변수가 모형에 추가된다고 하더라도 값이 커질 수 있다.  
* 수정 결정계수  
다중회귀분석(독립변수가 2개이상)을 수행하는 경우 결정계수의 한계를 보완하기 위해 표본의 크기(n)와 독립변수의 수(p)를 고려하여 수정된 결정계수를 구한다. $$Adj.R^2 = 1 - \frac{(n-1)}{(n-p-1)}(1-R^2)$$
>case)$$Adj.R^2 = 1 - \frac{(180-1)}{(180-3-1)}(1- 0.62) = 0.6135$$  


6. F-statistic & Prob (F-statistic) : 회귀모델 전체의 통계적 유의성 검정
$$y = \beta_0 + \beta_1x_0 + \beta_2x_1 +\epsilon$$

$$\begin{cases}
   H_0 : \beta_0 = \beta_1 = \beta_2 = 0 \\
   H_1 : 모든 \beta_i (i = 0,1,2) 가 \space 0은 아니다.
\end{cases}$$
>Case : Prob (F-statistic) 7.88e-37 < 0.05
>1. Decision : $H_0$ 기각한다
>2. Conclusion : 모델의 회귀계수 중 적어도 하나는 0이 아니다 -> 모델 전체의 회귀성 존재한다.

7. AIC & BIC
Data를 이용하여 모형만들 때, Overfit된 모형(복잡한 모형)을 만든다면, 새로운 Data에 대해 설명력이 떨어진다. 따라서 복잡한 모형에 대한 패널티를 부여하여 좀 더 Roubust한 모형을 선택하게 하는 기준이 되는 지표다.  더 작은 값을 가진 모형을 선택한다
* AIC : 실제 Data와 모형을 통해 예측한 Data 분포가 동일한지 판단 $$AIC = -2log(likelihood) + 2p$$
* BIC : 실제 Data 표본의 크기도 고려하여 분포의 동일성 판단 (표본이 큰 경우 사용) $$BIC = -2log(likelihood) + plog(n)$$
where, p = 변수 개수, n : 데이터 수
>Case : Constant 항이 있을 땐 변수에 + 1을 해준다
>* AIC = -2*(391.72) + 2*(3+1) = -775.44
>* BIC = -2*(391.72) + (3+1)*log(180) = -762.66

Case.2 : $\alpha = 0$
$$BRK-RF =  \beta_{MKT}(RM-RF) + \beta_{SMB}(SMB) + \beta_{HML}(HML) +\epsilon$$

In [16]:
Y12 = brk['BRK-RF']
X12 = brk[['MKT-RF', 'SMB', 'HML']]
model = sm.OLS(Y12,X12).fit()
print(model.summary())

                                 OLS Regression Results                                
Dep. Variable:                 BRK-RF   R-squared (uncentered):                   0.620
Model:                            OLS   Adj. R-squared (uncentered):              0.614
Method:                 Least Squares   F-statistic:                              96.25
Date:                Thu, 09 Mar 2023   Prob (F-statistic):                    5.48e-37
Time:                        12:30:23   Log-Likelihood:                          390.51
No. Observations:                 180   AIC:                                     -775.0
Df Residuals:                     177   BIC:                                     -765.4
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

## Problem 1-2

Case.1
$$
BRK-RF = \alpha + \beta_{MKT}(RM-RF) + \beta_{SMB}(SMB) + \beta_{HML}(HML) + \beta_{UMD}(UMD)+\epsilon
$$

In [18]:
Y21 = brk['BRK-RF']
X21 = brk[['MKT-RF', 'SMB', 'HML','UMD']]
X21 = sm.add_constant(X21, has_constant = 'add')
model = sm.OLS(Y21,X21).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 BRK-RF   R-squared:                       0.642
Model:                            OLS   Adj. R-squared:                  0.633
Method:                 Least Squares   F-statistic:                     78.35
Date:                Thu, 09 Mar 2023   Prob (F-statistic):           5.69e-38
Time:                        13:29:40   Log-Likelihood:                 396.89
No. Observations:                 180   AIC:                            -783.8
Df Residuals:                     175   BIC:                            -767.8
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0040      0.002      1.921      0.0

Case.2 : $\alpha$ = 0
$$
BRK-RF = \beta_{MKT}(RM-RF) + \beta_{SMB}(SMB) + \beta_{HML}(HML) + \beta_{UMD}(UMD)+\epsilon
$$

In [20]:
Y22 = brk['BRK-RF']
X22 = brk[['MKT-RF', 'SMB', 'HML','UMD']]
model = sm.OLS(Y22,X22).fit()
print(model.summary())

                                 OLS Regression Results                                
Dep. Variable:                 BRK-RF   R-squared (uncentered):                   0.639
Model:                            OLS   Adj. R-squared (uncentered):              0.630
Method:                 Least Squares   F-statistic:                              77.73
Date:                Thu, 09 Mar 2023   Prob (F-statistic):                    7.36e-38
Time:                        13:30:15   Log-Likelihood:                          395.01
No. Observations:                 180   AIC:                                     -782.0
Df Residuals:                     176   BIC:                                     -769.3
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

## Problem 1-3

Case.1
$$
BRK-RF = \alpha + \beta_{MKT}(RM-RF) + \beta_{SMB}(SMB) + \beta_{HML}(HML) + \beta_{QMJ}(QMJ) + \beta_{BAB}(BAB) + \beta_{UMD}(UMD)+\epsilon
$$

In [21]:
Y31 = brk['BRK-RF']
X31 = brk[['MKT-RF', 'SMB', 'HML','QMJ','BAB','UMD']]
X31 = sm.add_constant(X31, has_constant = 'add')
model = sm.OLS(Y31,X31).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 BRK-RF   R-squared:                       0.657
Model:                            OLS   Adj. R-squared:                  0.645
Method:                 Least Squares   F-statistic:                     55.20
Date:                Thu, 09 Mar 2023   Prob (F-statistic):           1.10e-37
Time:                        13:36:43   Log-Likelihood:                 400.79
No. Observations:                 180   AIC:                            -787.6
Df Residuals:                     173   BIC:                            -765.2
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0022      0.002      1.018      0.3

Case.2 : $\alpha $ = 0
$$
BRK-RF = \beta_{MKT}(RM-RF) + \beta_{SMB}(SMB) + \beta_{HML}(HML) + \beta_{QMJ}(QMJ) + \beta_{BAB}(BAB) + \beta_{UMD}(UMD)+\epsilon
$$

In [22]:
Y32 = brk['BRK-RF']
X32 = brk[['MKT-RF', 'SMB', 'HML','QMJ','BAB','UMD']]
model = sm.OLS(Y32,X32).fit()
print(model.summary())

                                 OLS Regression Results                                
Dep. Variable:                 BRK-RF   R-squared (uncentered):                   0.659
Model:                            OLS   Adj. R-squared (uncentered):              0.647
Method:                 Least Squares   F-statistic:                              56.04
Date:                Thu, 09 Mar 2023   Prob (F-statistic):                    3.86e-38
Time:                        13:37:17   Log-Likelihood:                          400.26
No. Observations:                 180   AIC:                                     -788.5
Df Residuals:                     174   BIC:                                     -769.4
Df Model:                           6                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

## INSIGHT

>1. Factor가 추가될 수록 Adj.$\space R^2$ 값이 증가한다; Berkshire Hathaway 초과수익률 변동성을 잘 포착한다(설명을 잘한다).(61.4% -> 64.7%)  
>2. $\alpha$ = 0인 경우 각 Factor에 대한 민감도가 증가한다(Coefficient 절대값이 증가한다).  
>3. 표본수(n = 180)을 고려하려 BIC값을 비교하면 $\alpha$ = 0 일때 상대적으로 실제 데이터 분포와 예측 데이터 분포가 더 비슷하다.(BIC크기가 상대적으로 작다)  
>4. Factor를 추가했을 때 드라마틱한 BIC변화는 보이지 않는다.